Create Monte Carlo Workspace Algorithm

Despiix opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
--> To benchmark the Mantid fitting engine, we need a method to generate randomly distributed data based on specific fit functions or workspaces. This will allow us to compare the generated data with known true values, providing insights into the performance of the fitting engine.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
--> I implemented a Mantid algorithm that uses a Cumulative Distribution Function (CDF) approach. This algorithm:

1. Takes a Workspace as Input: The input workspace defines the distribution’s characteristics.
2. Generates a Random Distribution: Using the CDF of the input data, the algorithm samples random points, producing a distribution that mirrors the input.
3. Ensures Reproducibility: The algorithm seeds the random number generator, ensuring reproducible results for consistent benchmarking.

Fixes #38196. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

**Report to:**  @thomashampson 

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
--> **Expected Outcome:**
This algorithm allows us to create random distributions based on known values, making it easier to assess the fitting engine’s accuracy and performance across different scenarios.

### To test:
<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
1. Validate my tests
2. Run the tests
3. Verify algorithm does what it's expected to do.
4. Verify that documentation is correct + Location of image is correct

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Despiix committed:
Created Files and draft comments

Despiix committed:
Functions to calculate the variance and the mean of the vector containing the random numbers generated.

Despiix committed:
Added the Probability Density Function

+ Changed outputWS values

Refactor: Clean up code and improve readability

Despiix committed:
Version 1 using Cumulative Distribution Function

Clean up the codebase and break code into smaller functions.

Remove auto and outputY from function Headers

Despiix committed:
Add 3 Unit tests

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Added 2 Unit tests to reproducability and out of bounds handling

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Despiix committed:
Add progress bar

Despiix committed:
Add Algorithm Documentation

Release Notes

Add Description and remove redundant 0s

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jhaigh0 commented on Framework/Algorithms/inc/MantidAlgorithms/CreateMonteCarloWorkspace.h:
```diff
@@ -0,0 +1,39 @@
+namespace Algorithms {
+using namespace std;
+
+/** CreateMonteCarloWorkspace : TODO: DESCRIPTION
+ */
```
It'd be good to add a description here

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,148 @@
+
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  return static_cast<int>(std::round(total_counts));
```
What if this rounded to 0? I suppose you'd get 0s in the data, this might be confusing? Perhaps at least throw a warning?
Would it make more sense to add the number of MC events as an input parameter? Say for example you want to see how counting times/stats affect optimiser performance, I think to do this you would need to scale the input workspace which seems a  bit clunky.

thomashampson replied:
I think having the parameter to specify number of MC events is good, and the default could be the integral of the input.

Despiix replied:
I commited the suggestion above, please have a look and let me know. The scaling function is not perfect, if the number of MC events is too little, the new workspace appears to be a flat line when overplotted onto the original workspace. I am open to any suggestions on ways to improve it.

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,148 @@
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  outputWs->mutableY(0) = outputY;
```
Would it also make sens e to set the error as the sqrt of the Y? The user could call `SetUncertainties` algorithm for this I suppose...

Despiix replied:
That is a good idea I will look into adding that!

RichardWaiteSTFC replied:
As discussed offline, is it possible to check (by running your simulation with different seeds) that the observed distribution of counts is Poisson distributed?

thomashampson replied:
I think setting the error for each bin to sqrt(N) is appropriate in this case. Good suggestion.

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  MatrixWorkspace_sptr outputWs = WorkspaceFactory::Instance().create(instWs);
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Maybe a stupid question, sorry, but why do you need these sleep statements?

jhaigh0 replied:
I think you might've have the progress bar a little wrong (although I'm no expert).

If your algorithm, for example, iterated over the spectra of an input workspace and performed some kind of operation with each, you would have the progress bar update after each spectra operation. I think that's the idea anyway, it's showing the progress of the main bulk of the computation. I think for you this would be in `fillHistogramWithRandomData`.

I think you can also set it to display text (e.g. "writing output"). Maybe you can have a look at some other algorithms and see how it's used there.

Despiix replied:
I updated the code for the progress bar, although personally I prefer the previous look, the new version does not need the sleep statements.

RichardWaiteSTFC reviewed (commented):
Sorry for sticking my oar in here - I found this very interesting (I do a similar thing to generate random data in 3D Q for single-crystal peak integration testing).

I'm aware it may be a work in progress/unfinished, so please forgive me, but I had a couple of questions:

(1) What is the advantage of this over using numpy random number generation like so (assuming the bin-width is small enough you don't need to worry about variation in the underlying pdf over the width of a bin) ?
```
params =  np.array([2.5e4, 0.06,  0.015, 30000, 30,  50])

peak_func = FunctionFactory.Instance().createPeakFunction("BackToBackExponential")
[peak_func.setParameter(ipar, params[ipar]) for ipar in range(peak_func.nParams())]
comp_func = FunctionWrapper(peak_func) + FlatBackground(A0=params[-1])

# simuate data
np.random.seed(1)
x = np.linspace(ptrue[3]-350, ptrue[3]+500, 61)
y = np.random.poisson(comp_func(x))
e = np.sqrt(y)
ws = CreateWorkspace(DataX=x, DataY=y, DataE=e)
```
![image](https://github.com/user-attachments/assets/484b5b48-7d53-4648-a24d-090ebf920e22)

(2) This seems to only support single histogram workspaces, but it doesn't check the number of histograms in the input workspace (unless I missed it?) - would it just make sense to generalise it and add a (parallelised) loop over histograms?

Despiix commented:
> Sorry for sticking my oar in here - I found this very interesting (I do a similar thing to generate random data in 3D Q for single-crystal peak integration testing).
> 
> I'm aware it may be a work in progress/unfinished, so please forgive me, but I had a couple of questions:
> 
> (1) What is the advantage of this over using numpy random number generation like so (assuming the bin-width is small enough you don't need to worry about variation in the underlying pdf over the width of a bin) ?
> 
> ```
> params =  np.array([2.5e4, 0.06,  0.015, 30000, 30,  50])
> 
> peak_func = FunctionFactory.Instance().createPeakFunction("BackToBackExponential")
> [peak_func.setParameter(ipar, params[ipar]) for ipar in range(peak_func.nParams())]
> comp_func = FunctionWrapper(peak_func) + FlatBackground(A0=params[-1])
> 
> # simuate data
> np.random.seed(1)
> x = np.linspace(ptrue[3]-350, ptrue[3]+500, 61)
> y = np.random.poisson(comp_func(x))
> e = np.sqrt(y)
> ws = CreateWorkspace(DataX=x, DataY=y, DataE=e)
> ```

Thank you for taking the time to review my algorithm.  To answer to your first question based on my understanding: 

The main advantage of my approach is that it doesn't assume any specific underlying functional form for the data. By constructing a cumulative distribution function (CDF) directly from the input data and sampling from it, we can capture the actual empirical distribution, including any irregularities or complexities that may not be well-represented by a predefined function. This is done to deal with data where the underlying distribution is unknown, complex, or doesn't fit standard models.

Please correct the above if I am wrong, to be honest I hadn't considered using numpy random number generation. Let me know if you still think it would be a better alternative.

jhaigh0 commented on docs/source/release/v6.12.0/Inelastic/Algorithms/New_features/38196_CreateMonteCarloWorkspace.rst:
```diff
@@ -0,0 +1 @@
+- New algorithm that creates a randomly simulated workspace by sampling from the probability distribution of input data.
```
This needs to name the new algorithm and link to it (have a look at some other release notes for examples)

jhaigh0 commented on docs/source/algorithms/CreateMonteCarloWorkspace-v1.rst:
```diff
@@ -0,0 +1,56 @@
+
+    fig.show()
+
+Output:
+    A histogram similar to the original but generated in a random way.
```
```suggestion
```
This output section is for checking the output of a script, I think if you describe what the script does above that's fine

jhaigh0 commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  std::mt19937 gen(seed_input);
+
+  Mantid::HistogramData::HistogramY outputY = fillHistogramWithRandomData(cdf, numIterations, gen);
```
I think it would be nicer if `fillHistogramsWithRandomData` took the seed as the input rather than a `std::mt19937`. There's no need to construct `gen` outside the method, since it's not sued after again after this point.

jhaigh0 commented on Framework/Algorithms/test/CreateMonteCarloWorkspaceTest.h:
```diff
@@ -0,0 +1,152 @@
+
+    Mantid::HistogramData::HistogramY outputY = alg.fillHistogramWithRandomData(cdf, numIterations, gen);
+
+    // The total counts should be less than numIterations because some random numbers are not counted
+    auto sumCounts = std::accumulate(outputY.begin(), outputY.end(), 0.0);
+    TS_ASSERT_LESS_THAN(sumCounts, numIterations);
```
do you know that this will always happen, is there a chance these two value could be equal?

RichardWaiteSTFC replied:
I think these are equal by design in this algorithm, this is in contrast to the method of randomly generating data I posted, where the total counts itself would be Poisson distributed (I believe, I should check...) - I've asked @Despiix to check the distribution of counts in each bin from multiple simulations are Poisson distributed (or close enough).

RichardWaiteSTFC replied:
> I think having the parameter to specify number of MC events is good, and the default could be the integral of the input.

@thomashampson - what about making the number of MC events equal to a random number drawn from Poisson distribution with mean of integral of the input? This might help the Poisson-ness of the stats in each bin (if that makes sense...)

Despiix committed:
Fix Progress Bar, Edit Release Note

Despiix committed:
Update Tests

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Update CreateMonteCarloWorkspace-v1.rst

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

Despiix committed:
Add Input for MC events and ScaleInputToMatchMCEvents and the errors are calculated as the sqrt of the counts.

Despiix committed:
Update Scaling Function

Does not work if MC events are much less than data points in original workspace

Updated Tests

Remove redundant code

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,188 @@
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
```
I'm wondering whether one could optionally output an event workspace (which is essentially what you are simulating), or produce an event workspace and then rebin to match the input workspace?

sf1919 commented:
Does this need to move to v6.13?

Despiix commented:
> Does this need to move to v6.13?

No, I am waiting for someone to review it. All the additional features will be added through a seperate pr.

RichardWaiteSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/CreateMonteCarloWorkspace.h:
```diff
@@ -0,0 +1,44 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
```
Probably 2025 now!

RichardWaiteSTFC commented on Framework/Algorithms/CMakeLists.txt:
```diff
@@ -234,16 +235,16 @@ set(SRC_FILES
+    src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp
+    src/PolarizationCorrections/PolarizationCorrectionsHelpers.cpp
     src/PolarizationCorrections/PolarizationEfficienciesWildes.cpp
+    src/PolarizationCorrections/PolarizerEfficiency.cpp
```
These changes look unrelated to the PR, do you need to rebase?

RichardWaiteSTFC commented on Framework/Algorithms/test/CreateMonteCarloWorkspaceTest.h:
```diff
@@ -0,0 +1,169 @@
+    TS_ASSERT_DELTA(totalScaledCounts, 20.0, 1e-6); // Verify the scaled sum matches targetMCEvents
+  }
+
+  MatrixWorkspace_sptr createInputWorkspace(int numBins, double initialValue) {
```
Looks like there are a few helper functions here, these are good but perhaps separate them from the tests at the bottom of the file? Makes it easier to read somehow (perhaps just personal preference)

RichardWaiteSTFC reviewed (changes requested):
Thanks for addressing some of the points, in particular setting the errors.

This works well and the tests are good (checking reproducibility with the seed, respects number of MC events etc.).

I checked with the test workspace from the code I posted earlier, 
![image](https://github.com/user-attachments/assets/a1ef7923-ef65-470a-a48e-b9d1f7d166bb)
I also checked that the variance of the simulated counts agrees with the average over ~5000 simulations so I think it does seem reasonable to set e = sqrt(y), at least for large counts. I haven't tested the validity of the Poisson distribution etc.
![image](https://github.com/user-attachments/assets/54471d7a-5bed-4a53-a81b-356af672c153)

There are some small changes requested: perhaps some unnecessary normalisation done if the monte-carlo events are user specified?

I think there are some additional features I would like (support for workspaces with more than 1 spectrum, perhaps event workspaces as well) but these can be addressed in a separate PR as and when it becomes useful.

Despiix committed:
Adjust Algorithm and tests based on review

Update Style of code and image

change copyright year to 2025

Rename Variables

Update progress bar

Update Typo

Remove repetitive description

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

Co-authored-by: RichardWaiteSTFC <55979119+RichardWaiteSTFC@users.noreply.github.com>

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,169 @@
+  int seed_input = getProperty("Seed");
+  int userMCEvents = getProperty("MonteCarloEvents");
+
+  const Mantid::HistogramData::HistogramY &originalYData = inputWs->y(0); // Counts in each bin
```
Think it's preferred to use `auto` where possible, here's one example where I think you could
```suggestion
  const auto &originalYData = inputWs->y(0); // Counts in each bin
```

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,169 @@
+ *  Determine how many iterations to use for MC sampling.
+ *  If userMCEvents > 0, use that directly; otherwise use the integral of the input data.
+ */
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData,
```
Stylistically I would prefer to have this `if (userMCEvents > 0) ` in the `exec` and then this method would be called `integrateYData` or similar. But I'm happy to merge as is!

Despiix replied:
This should also be fixed now

RichardWaiteSTFC reviewed (dismissed):
Thanks for the changes, in particular renaming the .png and removing the unnecessary additional normalisation when the user supplies the number of events. 

I know this PR has been going a while, I'm happy to approve now as I think this algorithm works well and is important to continue with fitbenchmarking work. There are a few nit-picking comments, but as there are a few other features I'd like added separately after this release I think these can be addressed later! What do you think @jclarkeSTFC, @thomashampson, @jhaigh0  and @sf1919 ?

RichardWaiteSTFC commented:
Oh perhaps just noticed the copyright is still 2024, is that correct or should it be 2025? Again v minor...

thomashampson reviewed (changes requested):
I think we would benefit from a better image for the documentation to showcase the algorithm.

RichardWaiteSTFC commented:
> I think we would benefit from a better image for the documentation to showcase the algorithm.

Feel free to use this if you want
![image](https://github.com/user-attachments/assets/a1df671b-2527-4b03-930d-1d664a7aed1a)
or this if you prefer smooth input data
![image](https://github.com/user-attachments/assets/fb25831b-0dc4-446e-bf7f-8dd377d5a755)

RichardWaiteSTFC commented on docs/source/algorithms/CreateMonteCarloWorkspace-v1.rst:
```diff
@@ -0,0 +1,55 @@
+    func = FunctionWrapper(FunctionFactory.createInitialized("name=BackToBackExponential,I=25000,A=0.06,B=0.015,X0=30000,S=30;name=FlatBackground,A0=50"))
+    # create input workspace
+    x = np.linspace(29650.0, 30500.0, 201)
+    y = func(x)
```
Sorry this script reproduces the other image I sent ( the smooth blue curve). If you want the noisy data then you need
`y = np.random.poisson(func(x))`

Despiix replied:
Oh, I thought I added the other image. I will fix it now.

RichardWaiteSTFC reviewed (changes requested):
Thanks for the changes, I've tested and looks good!
Still got copyright 2024 here 
https://github.com/mantidproject/mantid/blob/39598fab2d6fe6116479a3ee04fbb4e3a50ef53c/Framework/Algorithms/inc/MantidAlgorithms/CreateMonteCarloWorkspace.h#L3
But test file has 2025 in the .cpp file - I don't know which one is right but I think they should be the same!
Other than that happy to approve!

Despiix commented:
> Thanks for the changes, I've tested and looks good! Still got copyright 2024 here
> 
> https://github.com/mantidproject/mantid/blob/39598fab2d6fe6116479a3ee04fbb4e3a50ef53c/Framework/Algorithms/inc/MantidAlgorithms/CreateMonteCarloWorkspace.h#L3
> 
> 
> But test file has 2025 in the .cpp file - I don't know which one is right but I think they should be the same!
> Other than that happy to approve!

Apologies for that I must have missed it! It should be fixed now.

RichardWaiteSTFC reviewed (dismissed):
Great work well done!

sf1919 commented:
I think @thomashampson may have to gatekeep this one

thomashampson commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,163 @@
+  mustBePositive->setLower(0);
+
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input Workspace containing data to be fitted");
```
```suggestion
                  "Input Workspace containing data to be simulated");
```

thomashampson commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,163 @@
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input Workspace containing data to be fitted");
+  declareProperty("Seed", 32, mustBePositive,
+                  "Integer that initializes a random-number generator, good for reproducibility");
```
```suggestion
                  "Integer that seeds initialises the random-number generator, for reproducibility");
```
Mentioning the word "seed" I think gives a lot of clarification.

thomashampson commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,163 @@
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         int seed_input,
```
```suggestion
                                                                                         int seedInput,
```
camelCase is our C++ standard

thomashampson reviewed (commented):
The new image is much better and really demonstrates the usefulness of the algorithm. I've made a couple of small comments so far, but haven't finished reading the code yet. I'm just submitting them now so they don't vanish.

RichardWaiteSTFC commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,163 @@
+ *  If userMCEvents > 0, use that directly; otherwise use the integral of the input data.
+ */
+int CreateMonteCarloWorkspace::integrateYData(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
```
Think you missed a variable here when converting to camel case sorry for not picking up on that in my previous reviews (maybe I do too much in python!)
```suggestion
  double totalCounts = std::accumulate(yData.begin(), yData.end(), 0.0);
```

Despiix replied:
Thanks for letting me know. I fixed it. I also squashed a lot of the commits to make it easier to follow

RichardWaiteSTFC reviewed (dismissed):
Thanks for doc changes and variable re-naming,  sorry for not picking up on the lack of camel case variable names in my previous reviews (maybe I do too much in python!). Just one I think you missed but otherwise LGTM!

thomashampson commented on Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp:
```diff
@@ -0,0 +1,163 @@
+
+  int numIterations = (userMCEvents > 0) ? userMCEvents : integrateYData(originalYData);
+
+  API::Progress progress(this, 0.0, 1.0, 2);
```
```suggestion
  API::Progress progress(this, 0.0, 1.0, 101);
```
Setting the progress reporter to have 101 steps (one for the cdf creation, and 100 for the percentages of the MC generation steps), then it starts to work as expected. I tested this by setting MonteCarloEvents=10000 to make it take 10 seconds to generate.

thomashampson replied:
Alternatively (maybe a neater solution because you don't need to pass the reporter through to the function?), you can just create a new progress reporter inside the generating function and give it 100 steps there, that seems to work fine when I test it.

thomashampson commented on docs/source/algorithms/CreateMonteCarloWorkspace-v1.rst:
```diff
@@ -0,0 +1,55 @@
+
+Description
+-----------
+The algorithm generates a simulated workspace by sampling from the probability
+distribution of input data, useful for testing of fitting functions and modeling.
+By generating a simulated dataset that mirrors the probability
+distribution of existing data.
```
Similar to the class description it looks like there's repetition here.

thomashampson committed:
rename variable

thomashampson reviewed (approved):
This adds a really nice algorithm that will be useful for generating data to benchmark the fitting engine and study the fitting in depth. It will also allow users to generate Monte Carlo data based on any input workspace, which may be usful for testing and validation.

The use of stl features has made the code very concise and efficient.
